paymentapi/src/test/java/fi/hel/verkkokauppa/payment/api/PaymentAdminControllerUnitTest.java

```diff
@@ -91,4 +144,335 @@ public void savePaymentFilter() throws Exception {
         Assertions.assertEquals(expected.getValue(), paymentFilterDto.getValue());
     }
-}
\ No newline at end of file
+
+
+    /**
+     * It tests the create payment method endpoint.
+     */
+    @Test
+    public void whenCreatePaymentMethodWithValidDataThenReturnStatus201() throws Exception {
+        PaymentMethodDto paymentMethodDto = createTestPaymentMethodDto(GatewayEnum.OFFLINE);
+        PaymentMethodModel paymentMethodModel = mapper.convertValue(paymentMethodDto, PaymentMethodModel.class);
+
+        Mockito.when(paymentMethodService.createNewPaymentMethod(paymentMethodDto)).thenCallRealMethod();
+        Mockito.when(paymentMethodRepository.findByCode(paymentMethodDto.getCode())).thenReturn(Collections.emptyList());
+        Mockito.when(paymentMethodRepository.save(paymentMethodModel)).thenReturn(paymentMethodModel);
+        Mockito.when(paymentMethodMapper.fromDto(paymentMethodDto)).thenReturn(paymentMethodModel);
+        Mockito.when(paymentMethodMapper.toDto(paymentMethodModel)).thenReturn(paymentMethodDto);
+
+        MvcResult response = this.mockMvc.perform(
+                        post("/payment-admin/payment-method")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .accept(MediaType.APPLICATION_JSON)
+                                .content(mapper.writeValueAsString(paymentMethodDto))
+                )
+                .andDo(print())
+                .andExpect(status().is2xxSuccessful())
+                .andExpect(status().is(201))
+                .andReturn();
+        PaymentMethodDto responseDto = mapper.readValue(response.getResponse().getContentAsString(), PaymentMethodDto.class);
+        Assertions.assertNotNull(responseDto);
+        Assertions.assertEquals(responseDto, paymentMethodDto);
+    }
+
+    @Test
+    public void whenCreatePaymentMethodWithInvalidGatewayDataThenReturnStatus400() throws Exception {
+        JSONObject json = createTestPaymentMethodDtoRawJson("invalid");
+        this.mockMvc.perform(
+                        post("/payment-admin/payment-method")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(mapper.writeValueAsString(json.toString()))
+                )
+                .andDo(print())
+                .andExpect(status().is4xxClientError())
+                .andExpect(status().is(400));
+
+        JSONObject json2 = createTestPaymentMethodDtoRawJson("OFFLINE");
+        this.mockMvc.perform(
+                        post("/payment-admin/payment-method")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(mapper.writeValueAsString(json2.toString()))
+                )
+                .andDo(print())
+                .andExpect(status().is4xxClientError())
+                .andExpect(status().is(400));
+    }
```

Comment:
Should there be assertions to error text?

Reply:
The error is thrown by jackson while doing the deserialization of request body so it does not have a custom message to assert.